Python/chapter04/p01_route_between_nodes/miguelHx.py

```diff
@@ -0,0 +1,202 @@
+"""Python Version 3.9.2
+4.1 - Route Between Nodes:
+Given a directed graph, design an algorithm to
+find out whether there is a route between two nodes.
+"""
+import unittest
+
+from collections import deque
+from dataclasses import dataclass
+from typing import List, Deque
+
+
+@dataclass
+class Graph:
+    nodes: 'List[Node]'
+
+    def print_graph(self):
+        for node in self.nodes:
+            node.print_children()
+
+    def reset_visited(self):
+        for node in self.nodes:
+            node.visited = False
+
+
+@dataclass
+class Node:
+    id: int
+    children: 'List[Node]'
+    visited: bool = False
+
+    def add_child(self, *nodes: 'Node'):
+        for node in nodes:
+            self.children.append(node)
+
+    def print_children(self):
+        message = f'Adjacency list for node ({self.id}): '
+        for child in self.children:
+            message += '{}, '.format(child.id)
+        print(message)
```

Comment:
for performance reasons, it's better to avoid concatenating to a string in a loop:

```suggestion
        logging.debug('Adjacency list for node %s: %s', self.id, ', '.join(str(child.id) for child in self.children))
```

also, the [`logging`](https://docs.python.org/3/library/logging.html) module is preferred for this, because it's configurable.

Reply:
Right, I remember that conversation we had before about concatenating to a string in a loop.  .join is preferred
Committed your suggestion here:  2ef8112845a08a2c0c804b221b286f423c786197